include/crocoddyl/core/solvers/ipopt/multiple-shooting-nlp.hpp

```diff
@@ -0,0 +1,178 @@
+                        const Ipopt::Number *x,
+                        bool                 new_x,
+                        Ipopt::Number        obj_factor,
+                        Ipopt::Index         m,
+                        const Ipopt::Number *lambda,
+                        bool                 new_lambda,
+                        Ipopt::Index         nele_hess,
+                        Ipopt::Index        *iRow,
+                        Ipopt::Index        *jCol,
+                        Ipopt::Number       *values);
+
+    /** This method is called when the algorithm is complete so the TNLP can store/write the solution */
+    virtual void finalize_solution(Ipopt::SolverReturn               status,
+                                   Ipopt::Index                      n,
+                                   const Ipopt::Number              *x,
+                                   const Ipopt::Number              *z_L,
+                                   const Ipopt::Number              *z_U,
+                                   Ipopt::Index                      m,
+                                   const Ipopt::Number              *g,
+                                   const Ipopt::Number              *lambda,
+                                   Ipopt::Number                     obj_value,
+                                   const Ipopt::IpoptData           *ip_data,
+                                   Ipopt::IpoptCalculatedQuantities *ip_cq);
+    //@}
+
+    bool intermediate_callback(Ipopt::AlgorithmMode              mode,
+                               Ipopt::Index                      iter,
+                               Ipopt::Number                     obj_value,
+                               Ipopt::Number                     inf_pr,
+                               Ipopt::Number                     inf_du,
+                               Ipopt::Number                     mu,
+                               Ipopt::Number                     d_norm,
+                               Ipopt::Number                     regularization_size,
+                               Ipopt::Number                     alpha_du,
+                               Ipopt::Number                     alpha_pr,
+                               Ipopt::Index                      ls_trials,
+                               const Ipopt::IpoptData           *ip_data,
+                               Ipopt::IpoptCalculatedQuantities *ip_cq);
+
+    const std::size_t &get_nvar() const;
+
+    const std::vector<Eigen::VectorXd>                  &get_xs() const;
+    const std::vector<Eigen::VectorXd>                  &get_us() const;
+    const boost::shared_ptr<crocoddyl::ShootingProblem> &get_problem() const;
+
+    void set_xs(const std::vector<Eigen::VectorXd> &xs);
+    void set_us(const std::vector<Eigen::VectorXd> &us);
+
+    private:
+    boost::shared_ptr<crocoddyl::ShootingProblem> problem_;
+    boost::shared_ptr<crocoddyl::StateAbstract>   state_;
```

Comment:
You don't need to store another shared pointer of the state.
You could retrieve it from `problem_`.

Reply:
How could I do that? I don't see any getter to retrieve the state from problem. You mean through the `runningModels`?

What is the problem of having another shared pointer of the state? Having this variable inside the class simplifies things when calling to integrate or its Jacobians